### PR TITLE
Added support for pigpio backend.

### PR DESCRIPTION
**Pull request by sphh**
[pigpio](http://abyz.me.uk/rpi/pigpio/) is another GPIO library for the Raspberry Pi implementing a server-client architecture. This is a driver which connects to the pigpio daemon.

In addition it has PWM drivers for the LCD backlight and contrast. (This was the main reason to implement a driver for the pigpio daemon, because the RPi.GPIO PWM drivers resulted in a somewhat flickering backlight and if used for contrast, horizontal strips running from top to bottom of the LCD.)

### CODE REVIEW

**Comment by dbrgn** [comment]
Thanks for this, and sorry for forgetting to review :)

Would be great if you could also update the [docs](https://github.com/dbrgn/RPLCD/tree/master/docs). You can build the docs locally with `make html`.

Also, if you're tripping over the PEP8 checks on TravisCI, you can run those tests locally:

```
pip install -r requirements-dev.txt
pytest
```

File: RPLCD/pigpio.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+                       backlight_pwm=False, backlight_enabled=True,
+                       pin_contrast=None, contrast_mode='active_high',
+                       contrast_pwm=None, contrast=0.5,
+                       cols=20, rows=4, dotsize=8,
+                       charmap='A02',
+                       auto_linebreaks=True):
+        """
+        Character LCD controller.
+
+        The default pin numbers are based on the BCM numbering scheme!
```
Comment (dbrgn) [inline RPLCD/pigpio.py]: If I read pigpio docs correctly, then they support *only* BCM numbering, right?

File: lcdtest.py [hunk 2]
```diff
@@ -192,8 +228,43 @@ def options_pop(value, default=no_default):
...
+        bl = None if bl == 'None' else int(bl)
+
+        # Parse data pins into a list
+        pins_data = {}
+        pins_data = data.split(',')
+        # Convert data pins to int
+        pins_data = [int(pin) for pin in pins_data]
+        lcd = pigpio.CharLCD(pi,
+                             pin_rs=rs, pin_rw=rw, pin_e=e, pins_data=pins_data, pin_backlight=bl,
+                             numbering_mode=numbering_mode, cols=cols, rows=rows, charmap=charmap)
```
Comment (dbrgn) [inline lcdtest.py]: You can remove the `numbering_mode` param, it's unused in pgpio mode.

Comment (sphh) [reply lcdtest.py]: You are absolutely right in both cases. Thanks for spotting those mistakes. I just pushed the changes.

**Comment by sphh** [comment]
I will add it to the docs. Might need a bit of time.

Thanks for pointing out how to build them. That helps a lot.

**Comment by sphh** [comment]
Sorry for the double push. I just figured out what that error message means (and how to get rid of it):
```
WARNING: autodoc: failed to import class 'CharLCD' from module 'RPLCD.pigpio'; the following exception was raised:
No module named 'pigpio'
```

**Comment by dbrgn** [comment]
Ah, sorry, I forgot about the mocking :)

File: docs/getting_started.rst [hunk 3]
```diff
@@ -62,6 +62,20 @@ the `Adafruit tutorial
 <https://learn.adafruit.com/character-lcds/wiring-a-character-lcd>`_ to learn
 how to wire up these circuits.
 
+Via pigpio
+~~~~~~~~~~
+
+If you decide to use the ``pigpio`` library to control the LCD, follow the
+instructions set out above. Please keep in mind that the ``pigpio`` can only
+use the BOARD numbering scheme.
```
Comment (dbrgn) [inline docs/getting_started.rst]: I think you meant to write BCM, right? https://raspberrypi.stackexchange.com/questions/12966/what-is-the-difference-between-board-and-bcm-for-gpio-pin-numbering

Comment (sphh) [reply docs/getting_started.rst]: Yes, you are right. Thanks for spotting that mistake. It should be sorted in the most recent change.

**Comment by dbrgn** [comment]
Thank you, merged in 677d1aa99de69f6d3d2fd924f8dd0236121f691d!

**Comment by dbrgn** [comment]
@sphh can you confirm that the current master version works nicely with pigpio? Then I can do a new PyPI release.

**Comment by sphh** [comment]
I just tested it. It works - but only if you have `RPi.GPIO` installed.

If you don't have `RPi.GPIO` installed, `RPLCD.__init__.py` fails, because it tries to import `RPLCD.gpio` which again imports `RPi.GPIO`. I guess this requires a break with the backward compatibility built into `__init__.py` ... :-(

**Comment by dbrgn** [comment]
Ahh, I see. :(

I'd like to avoid breaking compatibility. RPi.GPIO is preinstalled on Raspbian, so in most cases this should not be a problem.

We could probably import the gpio backend only if `RPLCD.CharLCD` is initialized by using some metaclass magic, but that increases complexity. Not sure if it's worth it.

**Comment by sphh** [comment]
I guess if someone uses pigpio, the chance that they get rid of RPi.GPIO is high. Yes, that's a tough decision ... If you decide to break backward compatibility, than it might be an idea to get rid of the contextmanagers too. Sometimes radical measures are needed if they improve things substantially ;-)

BTW an easy workaround for users needing these features is to simply not updating the library. This release does not add bug fixes or improve the RPLCD.gpio or RPLCD.i2c backends, so there is no need to update, isn't it? [Just my thoughts.]

**Comment by dbrgn** [comment]
The reason for the backwards compatibility is that there are tutorials around for Raspberry Pi newbies that suggest to do `pip install RPLCD`. If people follow those tutorials, then a new release will break the tutorial and their project for them. (If this were a regular library for experienced Python programmers, then I would totally agree that a breaking release is fine if the version number is increased.)

I think I found a way to solve this:

```
# foo/__init__.py
class Foo:
    def __new__(cls, *args, **kwargs):
        from .bar import Bar
        return Bar(*args, **kwargs)
```

```
# foo/bar.py
print('IMPORTED')

class Bar:
    def __init__(self, *args, **kwargs):
        print('*args', args)
        print('*kwargs', kwargs)
    def aa(self):
        print('aa')
```

If you just import the `foo` module, nothing happens:

    $ python -c 'import foo'
    $ python -c 'from foo import Foo'

If you create an instance of `Foo` though, the `bar` module is imported too:

    $ python -c 'from foo import Foo; f = Foo()'
    IMPORTED
    *args ()
    *kwargs {}

...and the instance behaves like `Bar`:

    $ python -c 'from foo import Foo; f = Foo(1, b=2); f.aa()'
    IMPORTED
    *args (1,)
    *kwargs {'b': 2}
    aa

I'm a slight bit afraid that it will break something though, but I guess that could be handled by testing :)

**Comment by dbrgn** [comment]
By the way, a nice side effect of that is that instances of `Foo` are not actually wrappers around `Bar`, they *are* `Bar`:

```python
$ python -c 'from foo import Foo; f = Foo(); print(type(f))'
IMPORTED
*args ()
*kwargs {}
<class 'foo.bar.Bar'>
```

**Comment by sphh** [comment]
Chapeau, Danilo!!

**Comment by dbrgn** [comment]
PR: https://github.com/dbrgn/RPLCD/pull/78

Would be nice if you could test this too, @sphh :)

**Comment by sphh** [comment]
Yep. It works for me.

**Pull request closed without merging**